pkg/ipcache/metadata.go

```diff
@@ -327,43 +417,64 @@ func RemoveLabelsFromIPs(
 
 		triggerer.TriggerPolicyUpdates(false, "kube-apiserver identity updated by removal")
 	}
+	for ip, pair := range toReplace {
+		hIP, key := IPIdentityCache.getHostIPCache(ip)
+		meta := IPIdentityCache.getK8sMetadata(ip)
+		IPIdentityCache.deleteLocked(ip, pair.old.Source) // remove old entry and replace
+		if _, err := IPIdentityCache.upsertLocked(ip, hIP, key, meta, Identity{
+			ID:     pair.new.ID,
+			Source: pair.new.Source,
+		}); err != nil {
+			log.WithError(err).WithFields(logrus.Fields{
+				logfields.IPAddr:   ip,
+				logfields.Identity: pair.new,
+			}).Error("Failed to replace ipcache entry with new identity after label removal. Traffic may be disrupted.")
+		}
+	}
+}
+
+type idPair struct {
+	old, new Identity
 }
 
-// RemoveLabels removes the given labels association with the given prefix. The
-// leftover labels are returned, if any.
+// removeLabels removes the given labels association with the given prefix. The
+// leftover labels are returned, if any, and a boolean indicating whether the
+// caller must generate a new identity with the leftover labels. If the caller
+// must do so, then the following must occur *in order* to avoid drops:
+//   1) policy recalculation must be triggered
+//   2) new identity must have a new entry upserted into the IPCache
```

Comment:
Technically (1) should involve not just _triggering_ the policy recalculation, but _implementing_ the policy recalculation. If it's just a trigger, then (1) and (2) will happen concurrently and may still lead to packet drops due to mapping to a newly allocated identity before the policy is calculated.

(This may be a bigger rework to figure out how to achieve this reliably given the structure of the code today).

Reply:
Are you saying that we should clarify the godoc or that the code actually doesn't guarantee these two points and that it can cause drops due to mapping to a newly allocated identity before the policy is calculated?

Because I do think it provides that guarantee*.

\* This is under an asterisk because I've realized there's a very small window of possibility for drops between the recalculation and the ipcache mapping update, added it to https://github.com/cilium/cilium/issues/17962#issue-1060751671.

Reply:
Or are you actually pointing out this exact window for drops? I get the feeling that you are actually referring to something else, which I'm not quite clear on.

Reply:
Both :)

Breaking down a bit further, I see that there is more like three steps:

(1) Trigger policy updates
(1.5) Push those updates down into the datapath
(2) Push the new identity into the ipcache

If 1.5 is strictly before (2), then we are OK. My assertion above is that (1) does not synchronously provide (1.5), so it can technically lead to an order of (1) -> (2) -> (1.5). In this case, the gap between (2) plumbing the ipcache into the underlying maps and (1.5) plumbing the policymaps may cause policy drops.

Reply:
Joe and I went through this offline. The summary is that there's definitely a small window for drops, however this is not exclusive to the kube-apiserver policy matching feature. It is actually possible for any policy change that selects two IPs. See [`policyAdd()`](https://github.com/cilium/cilium/blob/v1.11.0/daemon/cmd/policy.go#L432) in the daemon.

For posterity, this can be fixed by making the "trigger" synchronous instead of async, because that would allow the ipcache entries that are added after the trigger to be present only when the endpoints have the correct identities plumbed into their policy maps.